src/utils/AuthProvider.tsx

```diff
@@ -0,0 +1,116 @@
+'use client';
+
+import { AuthError, AuthResponse, Session } from '@supabase/supabase-js';
+import { createContext, useContext, useEffect, useMemo, useState } from 'react';
+import { UUID } from 'crypto';
+import supabase from '@/api/supabase/createClient';
+
+// TODO: resolve inconsistent null/undefined types for empty values
+export interface AuthContextType {
+  session: Session | undefined;
+  userId: UUID | undefined;
+  userEmail: string | undefined;
+  signIn: (email: string, password: string) => Promise<AuthResponse>;
+  signOut: () => Promise<AuthError | null>;
+  signUp: (
+    email: string,
+    password: string,
+    options: object,
+  ) => Promise<AuthResponse>;
+}
+
+export const AuthContext = createContext<AuthContextType | undefined>(
+  undefined,
+);
+
+// TODO: we might need error handling here?
+export const useAuth = () => useContext(AuthContext);
+
+export default function AuthProvider({
+  children,
+}: {
+  children: React.ReactNode;
+}) {
+  const [session, setSession] = useState<Session | undefined>(undefined);
+  const [userId, setUserId] = useState<UUID | undefined>(undefined);
+  const [userEmail, setUserEmail] = useState<string | undefined>(undefined);
+
+  // set the session, userId, and userEmail
+  const setAll = (newSession: Session | undefined) => {
+    if (!newSession) return;
+    setSession(newSession);
+    if (newSession?.user?.id) {
+      const sessionUserId = newSession.user.id as UUID;
+      const sessionUserEmail = newSession.user.email;
+      setUserId(sessionUserId);
+      setUserEmail(sessionUserEmail);
+    }
+  };
+
+  // on page load, check if there's a session and set it
+  useEffect(() => {
+    supabase.auth.getSession().then(({ data: { session: newSession } }) => {
+      setAll(newSession != null ? newSession : undefined);
+    });
+
+    supabase.auth.onAuthStateChange((_event, newSession) => {
+      setAll(newSession != null ? newSession : undefined);
+    });
+  }, []);
+
+  // sign in and set the session, userId, and userEmail
+  const signIn = async (email: string, password: string) => {
+    const value = await supabase.auth.signInWithPassword({
+      email,
+      password,
+    }); // will trigger onAuthStateChange to update the session
+
+    setAll(value.data.session ? value.data.session : undefined);
+
+    return value;
+  };
+
+  // sign out and clear the session, userId, and userEmail
+  const signOut = async () => {
+    const { error } = await supabase.auth.signOut();
+
+    setSession(undefined);
+    setUserId(undefined);
+    setUserEmail(undefined);
+
+    return error;
+  };
+
+  // sign up and set the session, userId, and userEmail
+  const signUp = async (email: string, password: string, options: object) => {
```

Comment:
are there situations where we will be calling `signUp` outside of the current instance in the signup page, though? this feels like a concern only if we were to call the function in different instances in the future, but it's not clear to me where this might be. maybe to make this cleaner, i can get rid of the `options` arg entirely since idk that it has much utility lol

Reply:
I doubt it, but it's still good practice to have well-defined types. I can imagine a scenario where we may want to modify arguments to the `signUp` function but get mildly inconvenienced by the lack of IntelliSense. Which is the worst inconvenience to have if I'm being completely honest.

Anywho, if it helps, a quick dive into the node modules give us this declaration:
```Typescript
options?: {
  emailRedirectTo?: string;
  data?: object;
  captchaToken?: string;
}
```

But I'll leave the choice (and blame) of implementing this onto you